plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        ObjectMetadata metadata = getS3ObjectMetadata(path);
+        if (metadata == null) {
+            // the bucket root does not exist
+            if (path.getName().isEmpty()) {
+                return false;
+            }
+
+            // check if this path is a directory
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(1), false);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, this::keysFromListing));
+            if (!objectKeysIterator.hasNext()) {
+                return false;
             }
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            // the bucket root requires special handling
+            if (!path.getName().isEmpty()) {
+                MediaType metadataContentType = MediaType.parse(metadata.getContentType());
+                if (!metadataContentType.is(DIRECTORY_MEDIA_TYPE)) {
+                    Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(1), false);
+                    Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, this::keysFromListing));
+                    if (!objectKeysIterator.hasNext()) {
+                        // delete the object
+                        return deleteObject(keyFromPath(path));
+                    }
+
+                    // Delete the object which has a key corresponding to the prefix to be deleted
+                    deleteObject(keyFromPath(path));
```

Comment:
Shouldn't we call `deleteObject(keyFromPath(path))` when the content type is `application/x-directory` or when the object name ends with `_$folder$` ?